### DISCUSSION TITLE
error-stack: Use along with or replacing `thiserror`?

### DISCUSSION THREAD

**Discussion by dpc** [hashintel/hash#736] [error-stack (Help)] [answered]
I'm trying to understand how does `error-stack` fit with other error libraries. Currently there are two types of libraries; application error handling (boxed) like `anyhow` and `eyre` and library level (unboxed, enum-based) like `thiserror`.

Which layer does `error-stack` trying to fit (app, lib, both?) and why does it fit there?

**Comment by TimDiekmann** [comment, marked as answer]
Hi @dpc!

The main difference between library-type errors and application-type errors is how it's exposed. `thiserror` does not appear in the public API of a crate, which makes it incredibly useful for libraries (NB: `derive(Error)` from `thiserror` is not limited to `enum`s). Downstream crates depending on `thiserror` can use the returned error like any other `Error`. Application-type errors like `anyhow` or `eyre` *do* appear in the public API. Additionally, error types from application-type errors usually don't implement `Error` (but usually implement a way returning a `&dyn Error` like using `AsRef<&dyn Error>`).

`error_stack::Report` does not (and will not) implement `Error`. However, we plan to implement a way to convert `Report` into an `Error`. This is an important step so crates depending on libraries using `error-stack` can easily use the `Report` without the need of using `error-stack` themselves. However, crates that are using `error-stack` and depend on libraries using `error-stack` as well directly benefit from it and its features.

Using `error-stack` in a public API requires downstream crates to deal with `Report`. This means, that this crate is not suited as well for libraries as `thiserror`. Given the planned features, however, `error-stack` is suited for libraries and provides dependent crates a lot of information about the error, which is hard to encapsulate if not using types outside of `Error`.
For applications, `error-stack` was designed to provide useful error messages, thus making this crate a great fit.

`error-stack` relies on a `Context` trait to change the context. A blanket implementation for `Error` is provided, so `thiserror` is an easy way to create a `Context`. `error-stack` *can* be built on top of any existing `Error` type.
